crates/ggml/README.md

```diff
@@ -0,0 +1,139 @@
+to support rapid development without sacrificing backwards-compatibility, GGML
+uses versioning to introduce improvements that may change the format of the
+encoding. For example, newer versions of GGML make use of
+[vocabulary](#vocabulary)-scoring, which introduces extra information into the
+encoding, as well as [mmap](https://en.wikipedia.org/wiki/Mmap), which enhances
+performance through memory-mapping. The first value that is present in a valid
+GGML file is a "magic number" that indicates the GGML version that was used to
+encode the model.
+
+### Hyperparamaters
+
+The term
+"[hyperparameter](<https://en.wikipedia.org/wiki/Hyperparameter_(machine_learning)>)"
+describes a value that is used to configure the behavior of a large language
+model; this is in contrast to the model's **parameters**, which are the
+[weights](#weights) that were derived in the training process that was used to
+create the model. Each model defines its own hyperparameter structure that
+defines the hyperparameter values accepted by that model. Valid GGML files must
+list these values in the correct order, and each value must be represented using
+the correct data type. Although hyperparameters are different across models,
+some attributes appear in the hyperparameters for most models:
+
+- `n_vocab`: the size of the model's [vocabulary](#vocabulary)
+- `n_embd`: the size of the model's
+  "[embedding](https://en.wikipedia.org/wiki/Word_embedding) layer", which is
+  used during prompt ingestion
+- `n_layer`: the number of layers in the model; each layer represents a set of
+  [weights](#weights).
+
+### Vocabulary
+
+As the name implies, a model's vocabulary comprises components that are used by
+the model to generate language (text). However, unlike the vocabulary of a
+human, which consists of _words_, the vocabulary of a large language model
+consists of "tokens". A token _can_ be an entire word, but oftentimes they are
+word _fragments_. Just like humans can compose millions of words from just a
+dozen or two letters, large language models use _tokens_ to express a large
+number of words from a relatively smaller number of components. Consider a
+vocabulary with the following tokens: `whi`, `ch` `le`, `who`, and `a`; this
+vocabulary can be used to create the English words "which", "while", "who", "a",
+and "leach". How would the behavior change if the model contained the following
+tokens: `wh`, `ich`, `ile`, `o`, and `leach`? Choices such as these allow
+model-creators to tune the behavior and performance of their models.
+
+As described above, the model's [hyperparameters](#hyperparamaters) typically
+contains a value that specifies the number of tokens in the vocabulary. The
+vocabulary is encoded as a list of tokens, each of which includes an unsigned
+32-bit integer that specifies the length of the token. Depending on the GGML
+version, the token may also include a 32-bit floating point score.
+
+[comment]: <> (I need help describing token-scoring)
```

Comment:
I'd suggest linking to this: https://huggingface.co/learn/nlp-course/chapter6/5

but essentially, each token is given a score / probability, and the tokenization algorithm tries to select for tokens with the maximal probability.